src/Microsoft.ML.Core/Utilities/Stream.cs

```diff
@@ -821,49 +669,59 @@ public static byte[] ReadByteArray(this BinaryReader reader, int size)
 
         public static BitArray ReadBitArray(this BinaryReader reader)
         {
-            int numBits = reader.ReadInt32();
+            int numBits = reader.ReadInt32(); // reading trusted capacity value from data stream
             Contracts.CheckDecode(numBits >= 0);
             if (numBits == 0)
                 return null;
-            var numBytes = (numBits + 7) / 8;
+            var numBytes = (numBits + 7) / 8; // trusted capacity value expected not to integer overflow
             var bytes = reader.ReadByteArray(numBytes);
             var returnArray = new BitArray(bytes);
             returnArray.Length = numBits;
             return returnArray;
         }
 
-        public static unsafe void ReadBytes(this BinaryReader reader, void* destination, long destinationSizeInBytes, long bytesToRead, ref byte[] work)
+        private static unsafe void ReadBinaryDataIntoSpan<T>(BinaryReader reader, Span<T> destination) where T : unmanaged
         {
             Contracts.AssertValue(reader);
-            Contracts.Assert(bytesToRead >= 0);
-            Contracts.Assert(destinationSizeInBytes >= bytesToRead);
-            Contracts.Assert(destination != null);
-            Contracts.AssertValueOrNull(work);
+            Contracts.Assert(!destination.IsEmpty);
+
+            // There are two considerations here. First, we want to keep all temporary arrays (even pooled arrays)
+            // under some threshold size. Second, when we project the Span<T> to bytes, we need to do it in chunks,
+            // as trying to project the entire span at once will lead to integer overflow if the byte length
+            // exceeds int.MaxLength.
 
-            // Size our read buffer to 70KB to stay off the LOH.
-            const int blockSize = 70 * 1024;
-            int desiredWorkSize = (int)Math.Min(blockSize, bytesToRead);
-            EnsureSize(ref work, desiredWorkSize);
+            const int maxChunkSizeInBytes = 70 * 1024;
+            int maxChunkSizeInElements = maxChunkSizeInBytes / sizeof(T);
+            Contracts.Assert(maxChunkSizeInElements > 0, "Unexpectedly large T.");
 
-            fixed (void* src = work)
+            // Rent a byte[] instead of a T[] to allow reuse of buffers across different types T.
+            byte[] rentedArray = ArrayPool<byte>.Shared.Rent(maxChunkSizeInElements * sizeof(T));
```

Comment:
>maxChunkSizeInElements * sizeof(T) [](http://example.com/codeflow?start=61&length=34)

isn't the `maxChunkSizeInBytes`?

Reply:
`maxChunkSizeInBytes` could technically be a larger value than `maxChunkSizeInElements * sizeof(T)`, but maybe it's best just to go for simplicity here. One sec and I'll push the new commit.